REGRESSION (278276@main): Crash when collecting Transparent DocumentMarkers.

megangardner opened the pull request:
#### 585ea5ce731bd4f65e124ce2e0271a78796413dd
<pre>
REGRESSION (278276@main): Crash when collecting Transparent DocumentMarkers.
<a href="https://bugs.webkit.org/show_bug.cgi?id=273710">https://bugs.webkit.org/show_bug.cgi?id=273710</a>
<a href="https://rdar.apple.com/127511853">rdar://127511853</a>

Reviewed by Wenson Hsieh.

Original implementation of transparent document markers
was incorrect and resulted in a WebProcess crash. They
need to be closer to draggedContent markers (which I hope
to combine them with someday) and be gathered and added
separately so they can work with selection.

* Source/WebCore/dom/DocumentMarkerController.cpp:
(WebCore::shouldInsertAsSeparateMarker):
* Source/WebCore/rendering/MarkedText.cpp:
(WebCore::MarkedText::collectForDocumentMarkers):
(WebCore::MarkedText::collectForDraggedAndTransparentType):
(WebCore::MarkedText::collectForDraggedContent): Deleted.
* Source/WebCore/rendering/MarkedText.h:
* Source/WebCore/rendering/RenderText.cpp:
(WebCore::RenderText::contentRangesBetweenOffsetsForType const):
(WebCore::RenderText::draggedContentRangesBetweenOffsets const): Deleted.
* Source/WebCore/rendering/RenderText.h:
* Source/WebCore/rendering/TextBoxPainter.cpp:
(WebCore::TextBoxPainter&lt;TextBoxPath&gt;::paintForegroundAndDecorations):
(WebCore::TextBoxPainter&lt;TextBoxPath&gt;::paintForeground):

Canonical link: <a href="https://commits.webkit.org/278363@main">https://commits.webkit.org/278363@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/4095d2cfb9974c07f1394363eeab1e261631cb66

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/50287 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/29580 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/2584 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/53546 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/977 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/52590 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/35808 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/624 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/41027 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/52386 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/27242 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/43283 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/22129 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/24654 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/536 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/8665 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/46639 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/600 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/55132 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/25383 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/62/builds/535 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/48431 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/26646 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/43466 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/47459 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/11039 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/27508 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/26376 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

rr-codes commented on Source/WebCore/rendering/MarkedText.cpp:
```diff
@@ -323,12 +322,22 @@ Vector<MarkedText> MarkedText::collectForDocumentMarkers(const RenderText& rende
+    else {
+        ASSERT_NOT_REACHED();
+        return { };
+    }
```
I think this should be refactored into a lambda which returns an optional `Type` and uses a switch-case

rr-codes commented on Source/WebCore/rendering/RenderText.cpp:
```diff
@@ -1431,23 +1431,23 @@ Vector<std::pair<unsigned, unsigned>> RenderText::draggedContentRangesBetweenOff
+        std::pair<unsigned, unsigned> contentRange;
+        contentRange.first = markerStart;
+        contentRange.second = markerEnd;
+        contentRanges.append(contentRange);
```
```suggestion
        contentRanges.append({ markerStart, markerEnd });
```

rr-codes commented on Source/WebCore/rendering/RenderText.h:
```diff
@@ -183,7 +184,7 @@ class RenderText : public RenderObject {
     bool containsOnlyCSSWhitespace(unsigned from, unsigned length) const;
 
-    Vector<std::pair<unsigned, unsigned>> draggedContentRangesBetweenOffsets(unsigned startOffset, unsigned endOffset) const;
+    Vector<std::pair<unsigned, unsigned>> contentRangesBetweenOffsetsForType(const DocumentMarker::Type, unsigned startOffset, unsigned endOffset) const;
```
no need for `const`

rr-codes commented on Source/WebCore/rendering/RenderText.cpp:
```diff
@@ -1422,7 +1422,7 @@ bool RenderText::containsOnlyCSSWhitespace(unsigned from, unsigned length) const
 }
 
-Vector<std::pair<unsigned, unsigned>> RenderText::draggedContentRangesBetweenOffsets(unsigned startOffset, unsigned endOffset) const
+Vector<std::pair<unsigned, unsigned>> RenderText::contentRangesBetweenOffsetsForType(const DocumentMarker::Type type, unsigned startOffset, unsigned endOffset) const
```
no need for `const`

whsieh commented on Source/WebCore/rendering/TextBoxPainter.cpp:
```diff
@@ -329,12 +329,15 @@ void TextBoxPainter<TextBoxPath>::paintForegroundAndDecorations()
             }
+            auto markedTextsForTransparentContent = MarkedText::collectForDraggedAndTransparentType(DocumentMarker::Type::TransparentContent, m_renderer, m_selectableRange);
+            if (!markedTextsForTransparentContent.isEmpty())
+                markedTexts.appendVector(markedTextsForTransparentContent);
```
Nit -

```suggestion
                markedTexts.appendVector(WTFMove(markedTextsForTransparentContent));
```

(We could also fix the existing code above)

whsieh commented on Source/WebCore/rendering/TextBoxPainter.cpp:
```diff
@@ -516,6 +519,7 @@ void TextBoxPainter<TextBoxPath>::paintForeground(const StyledMarkedText& marked
         textPainter.setShadow(debugShadow);
 
     bool isTransparentMarkedText = markedText.type == MarkedText::Type::DraggedContent || markedText.type == MarkedText::Type::TransparentContent;
+
```
Nit - stray newline.

whsieh commented on Source/WebCore/rendering/TextBoxPainter.cpp:
```diff
@@ -329,12 +329,15 @@ void TextBoxPainter<TextBoxPath>::paintForegroundAndDecorations()
             bool shouldPaintDraggedContent = !(m_paintInfo.paintBehavior.contains(PaintBehavior::ExcludeSelection));
             if (shouldPaintDraggedContent) {
-                auto markedTextsForDraggedContent = MarkedText::collectForDraggedContent(m_renderer, m_selectableRange);
+                auto markedTextsForDraggedContent = MarkedText::collectForDraggedAndTransparentType(DocumentMarker::Type::DraggedContent, m_renderer, m_selectableRange);
```
Nit - I think `collectForDraggedAndTransparentContent` might be a slightly more accurate name.

whsieh reviewed (approved):
r=me, with the typo in the title fixed.

megangardner committed:
REGRESSION (278276@main): Crash when collecting Transparent DocumentMarkers.
https://bugs.webkit.org/show_bug.cgi?id=273710
rdar://127511853

Reviewed by Wenson Hsieh.

Original implementation of transparent document markers
was incorrect and resulted in a WebProcess crash. They
need to be closer to draggedContent markers (which I hope
to combine them with someday) and be gathered and added
separately so they can work with selection.

* Source/WebCore/dom/DocumentMarkerController.cpp:
(WebCore::shouldInsertAsSeparateMarker):
* Source/WebCore/rendering/MarkedText.cpp:
(WebCore::MarkedText::collectForDocumentMarkers):
(WebCore::MarkedText::collectForDraggedAndTransparentType):
(WebCore::MarkedText::collectForDraggedContent): Deleted.
* Source/WebCore/rendering/MarkedText.h:
* Source/WebCore/rendering/RenderText.cpp:
(WebCore::RenderText::contentRangesBetweenOffsetsForType const):
(WebCore::RenderText::draggedContentRangesBetweenOffsets const): Deleted.
* Source/WebCore/rendering/RenderText.h:
* Source/WebCore/rendering/TextBoxPainter.cpp:
(WebCore::TextBoxPainter<TextBoxPath>::paintForegroundAndDecorations):
(WebCore::TextBoxPainter<TextBoxPath>::paintForeground):

Canonical link: https://commits.webkit.org/278363@main
